### PR TITLE
JAMES-3295 Matcher for permanent delivery error

### PR DESCRIPTION
**Pull request by LanKhuat**

### CODE REVIEW

File: server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentError.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * Example:
+ *
+ * <pre><code>
+ * &lt;mailet match=&quot;IsRemoteDeliveryPermanentError&quot; class=&quot;&lt;any-class&gt;&quot;/&gt;
+ * </code></pre>
+ */
+public class IsRemoteDeliveryPermanentError extends GenericMatcher {
+
+    public static final AttributeName IS_DELIVERY_PERMANENT_ERROR = AttributeName.of("is-delivery-permanent-error");
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentError.java]: Please reuse the error code attribute added in https://github.com/linagora/james-project/pull/3552 (delivery-error-code), don't add yet-another attribute!

Comment (Arsnael) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentError.java]: You need to check in the mailet if the code is a permanent error or a temporary one right? What codes are considered permanent then?

Comment (LanKhuat) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentError.java]: or we will define another attribute for temporary errors? I don't really get this...

Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentError.java]: To be explicit 5xx -> permanent, 4xx -> we can retry

Souce https://www.wikiwand.com/en/List_of_SMTP_server_return_codes#/4xx_Persistent_transient_failure

**Comment by chibenwa** [comment]
`RabbitMQDeletedMessageVaultIntegrationTest>DeletedMessageVaultIntegrationTest.vaultEndpointShouldRestoreImapDeletedMailbox:255 » ConditionTimeout`

File: server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentError.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * <pre><code>
+ * &lt;mailet match=&quot;IsRemoteDeliveryPermanentError&quot; class=&quot;&lt;any-class&gt;&quot;/&gt;
+ * </code></pre>
+ */
+public class IsRemoteDeliveryPermanentError extends GenericMatcher {
+
+    @Override
+    public Collection<MailAddress> match(Mail mail) {
+        return mail.getAttribute(IS_DELIVERY_PERMANENT_ERROR)
+            .map(any -> mail.getRecipients())
```
Comment (Arsnael) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentError.java:48]: With what we said this morning you probably need to do a filter to check if it's true then

File: server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentError.java (hunk 1 above)
Comment (LanKhuat) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentError.java]: Permanent error could include other kinds of error besides SMTP, so adding another attribute would be more convenient.

**Comment by chibenwa** [comment]
test this please

File: server/mailet/mailets/src/test/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentErrorTest.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        testee.init(matcherConfig);
+    }
+
+    @Test
+    void shouldMatchWhenAttributeIsPresent() throws Exception {
+        Mail mail = createMail();
+        mail.setAttribute(new Attribute(IS_DELIVERY_PERMANENT_ERROR, AttributeValue.of(true)));
+
+        Collection<MailAddress> actual = testee.match(mail);
```
Comment (blackheaven) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/matchers/IsRemoteDeliveryPermanentErrorTest.java:68]: can be inlined (as the followings)

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**